### PR TITLE
Update ghc-9.0.1 support

### PR DESCRIPTION
**Pull request by jneira**
* Update stack dependencies and add ourmolu, floskell formatters and class, refine imports plugins
* Enable shake-bench for stack and ghc-9.0.1
* ~~Use hie-bios master~~ Use last hie-bios
* Clean up cabal-ghc901.project
* Make progress towards #297

### CODE REVIEW

File: cabal-ghc901.project [hunk 1]
```diff
@@ -24,6 +24,9 @@ packages:
          ./plugins/hls-module-name-plugin
          ./plugins/hls-ormolu-plugin
          ./plugins/hls-call-hierarchy-plugin
+
+with-compiler: ghc-9.0.1
+
```
Comment (pepeiborra) [inline cabal-ghc901.project:29]: what does this do?

File: cabal-ghc901.project [hunk 2]
```diff
@@ -45,9 +48,9 @@ source-repository-package
 
 source-repository-package
   type: git
-  location: https://github.com/jneira/hie-bios/
-  tag: 9b1445ab5efcabfad54043fc9b8e50e9d8c5bbf3
-  -- https://github.com/mpickering/hie-bios/pull/285
+  location: https://github.com/mpickering/hie-bios/
+  tag: 1875bff093983a0506f80e214eda27e7419da3bc
+  -- https://github.com/mpickering/hie-bios/pull/300
```
Comment (pepeiborra) [inline cabal-ghc901.project:53]: is there an upstream issue to upload this to Hackage?

**Review by pepeiborra — APPROVED** [review]

File: cabal-ghc901.project (hunk 1 above)
Comment (jneira) [reply cabal-ghc901.project:29]: It is the equivalent to do `-w ghc-9.0.1` or `--with-compiler ghc-9.0.1`, make the build use that compiler executable

File: cabal-ghc901.project (hunk 2 above)
Comment (jneira) [reply cabal-ghc901.project:53]: I've commented in the mentioned pr about: https://github.com/mpickering/hie-bios/pull/300#issuecomment-906417219

Comment (fendor) [reply cabal-ghc901.project:53]: Release is done

**Comment by jneira** [comment]
[the stack build](https://app.circleci.com/pipelines/github/jneira/haskell-language-server/704/workflows/9c902c6f-a6eb-4cf5-8232-6e5321d7fd80/jobs/6066) is failing with
```
hls-refine-imports-plugin        > /root/build/plugins/hls-refine-imports-plugin/src/Ide/Plugin/RefineImports.hs:52:1: error:
hls-refine-imports-plugin        >     Could not load module ‘TcRnMonad’
hls-refine-imports-plugin        >     It is a member of the hidden package ‘ghc-9.0.1’.
hls-refine-imports-plugin        >     Perhaps you need to add ‘ghc’ to the build-depends in your .cabal file.
hls-refine-imports-plugin        >     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
hls-refine-imports-plugin        >    |
hls-refine-imports-plugin        > 52 | import           TcRnMonad                            (initTcWithGbl,
hls-refine-imports-plugin        >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

--  While building package hls-refine-imports-plugin-1.0.0.1 (scroll up to its section to see the error) using:
      /root/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.4.0.0_ghc-9.0.1 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.4.0.0 build lib:hls-refine-imports-plugin --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1

Exited with code exit status 1

CircleCI received exit code 1
```

but the .cabal file has ghc as dependency:

https://github.com/haskell/haskell-language-server/blob/a800b9d71e4eac70d672d30b2488949ea26dcbc3/plugins/hls-refine-imports-plugin/hls-refine-imports-plugin.cabal#L24

**Comment by fendor** [comment]
> [the stack build](https://app.circleci.com/pipelines/github/jneira/haskell-language-server/704/workflows/9c902c6f-a6eb-4cf5-8232-6e5321d7fd80/jobs/6066) is failing with
> 
> ```
> hls-refine-imports-plugin        > /root/build/plugins/hls-refine-imports-plugin/src/Ide/Plugin/RefineImports.hs:52:1: error:
> hls-refine-imports-plugin        >     Could not load module ‘TcRnMonad’
> hls-refine-imports-plugin        >     It is a member of the hidden package ‘ghc-9.0.1’.
> hls-refine-imports-plugin        >     Perhaps you need to add ‘ghc’ to the build-depends in your .cabal file.
> hls-refine-imports-plugin        >     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
> hls-refine-imports-plugin        >    |
> hls-refine-imports-plugin        > 52 | import           TcRnMonad                            (initTcWithGbl,
> hls-refine-imports-plugin        >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
> 
> --  While building package hls-refine-imports-plugin-1.0.0.1 (scroll up to its section to see the error) using:
>       /root/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.4.0.0_ghc-9.0.1 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.4.0.0 build lib:hls-refine-imports-plugin --ghc-options " -fdiagnostics-color=always"
>     Process exited with code: ExitFailure 1
> 
> Exited with code exit status 1
> 
> CircleCI received exit code 1
> ```
> 
> but the .cabal file has ghc as dependency:
> 
> https://github.com/haskell/haskell-language-server/blob/a800b9d71e4eac70d672d30b2488949ea26dcbc3/plugins/hls-refine-imports-plugin/hls-refine-imports-plugin.cabal#L16-L31

Looks like this is not depending on ghc-api-compat, and ghc 9.0 does not expose `TcRnMonad` anymore

**Comment by jneira** [comment]
refineImports was disabled by flags for cabal (but the package was not 🤦 )
well as @fendor has got to compile all of them i am gonna touch the less possible

**Comment by jneira** [comment]
Now it is failing with:

```
src/Control/Concurrent/Strict.hs:2:6: error:
Error:     Ambiguous occurrence ‘modifyVar'’
    It could refer to
       either ‘Control.Concurrent.Extra.modifyVar'’,
              imported from ‘Control.Concurrent.Extra’ at src/Control/Concurrent/Strict.hs:7:1-72
           or ‘Control.Concurrent.Strict.modifyVar'’,
              defined at src/Control/Concurrent/Strict.hs:15:1
  |
2 |     (modifyVar', modifyVarIO'
  |      ^^^^^^^^^^
```

I think it is due the recent release of the extra library, aiiiins

**Comment by jneira** [comment]
I am afraid we cant use hlint-3.2.* and the last extra lib version (@ndmitchell set a upper bound to exclude it https://hackage.haskell.org/package/hlint-3.2.7/revision/1.cabal)

```
 [42 of 59] Compiling Hint.Fixities    ( src/Hint/Fixities.hs, dist/build/Hint/Fixities.o, dist/build/Hint/Fixities.dyn_o )

src/Hint/Fixities.hs:65:38: error:
Error:     Ambiguous occurrence ‘!?’
    It could refer to
       either ‘Data.List.Extra.!?’,
              imported from ‘Data.List.Extra’ at src/Hint/Fixities.hs:22:1-22
           or ‘Data.Map.!?’,
              imported from ‘Data.Map’ at src/Hint/Fixities.hs:23:1-15
              (and originally defined in ‘Data.Map.Internal’)
   |
65 |     case compareFixity <$> (fixities !? occNameString lop) <*> (fixities !? occNameString rop) of
   |        
```

**Comment by jneira** [comment]
i know add new definitions does not suppose bump up the breaking api version following pvp but if they clash with widely used defs as `Data.Map`, i think it had to have been `extra-1.8` imho

**Comment by ndmitchell** [comment]
@jneira - I think you're right - had I realised that extra was going to break as many things as it did, I'd have definitely made it 1.8. However, I was completely unaware in advance, and now the choice is whether to yank 1.7.10 and upload 1.8, and I suspect at that now that's just adding more hassle? But happy to be convinced otherwise. Sorry!

**Comment by jneira** [comment]
no worries, it is hard to know in advance where defs are being shadowed if the lib itself does no use them
tbh if client code uses qualified imports adding new defs should not break panything
yeah I think it can be kept as is, if a serious bug in hlint needs to be backported a new release can be still done

**Pull request merged**